### ISSUE TITLE
Implement a high-performance ring buffer for RTS built using CAS primitives.

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz-zio#35]
The ring buffer must provide support for a limited number of readers (equal to thread count), and potentially a higher number of writers. That is, many places in the code will add work to the queue, but few will read them. 

In addition, metrics are important to create good preemption. Metrics on how many ops per second, metrics on how many work units per queue, and so forth, can all be good to feed into an algorithm that decides whether or not a current thread should cooperatively yield to another fiber.

**Comment by regiskuckaertz** [comment]
This question is completely unrelated, but in `IOQueue` you wrote:

> Investigate using a faster option than `Queue`, because `Queue` has `O(n)` `length` method

I was wondering if a ring buffer would be a better implementation for the queue itself? Provided the invariant encoded by the `State` ADT is maintained.

It would probably require that _0 <= size <= capacity_ though, which to me looks reasonable 🤔 Everything below 0 or over capacity represents a notion of "potential" for what the queue size will be in the future.

**Comment by jdegoes** [comment]
We should work with #36 on this issue. A key part will be defining the "thread pool" like abstraction which is used to provide this functionality.

```scala
private[zio] trait ZioThreadPool extends ThreadPool {
  def concurrency: Int
  def capacity: Int 
  
  def size: Int
  def executions: Long
  def duration: Duration

  def submit(thunk: () => Unit): Unit
  def shutdown: Unit
  def shutdownNow: Unit
}
```

**Comment by artempyanykh** [comment]
Here's a high-level description of the RingBuffer project: https://gist.github.com/artempyanykh/d6a6e7a305b036727bf7995bf102b4ca

I've found these materials to be relevant, might be a good starting point:

1. Article: [The LMAX Architecture](https://martinfowler.com/articles/lmax.html) by Martin Fowler.
2. Article: [Yet another implementation of a lock-free circular array queue](https://www.codeproject.com/Articles/153898/Yet-another-implementation-of-a-lock-free-circular) by Faustino Frechilla is a deep dive into the subject with examples in C++.
3. Video: [LMAX - How to do 100K TPS at less than 1ms latency](https://www.infoq.com/presentations/LMAX) by Martin Thompson is the first LMAX talk by one of the authors of LMAX.
4. Video: [Locks? We Don't Need No Stinkin' Locks!](https://www.youtube.com/watch?v=VBnLW9mKMh4) by Michael Barker is about v3 of LMAX that received a proper MPMC support.
5. Paper: [Draft: Lock-Free Ringbuffer Designs](https://pdfs.semanticscholar.org/8c15/b938e6fdbcbab6e1b5704857043e5c680734.pdf).
6. Paper: [Simple, Fast, and Practical Non-Blocking and Blocking Concurrent Queue Algorithms](http://www.cs.rochester.edu/~scott/papers/1996_PODC_queues.pdf)

**Comment by artempyanykh** [comment]
I did some additional reading and research on the subject. Below are some of my current thoughts.

**TL;DR**

The interface for a ring buffer:

```scala
abstract class MPMCQueue(requestedCapacity: Int) {
  val capacity: Int
  def relaxedSize(): Int // not sure if we need accurate size, probably some lag (+/-1) is acceptable
  def processedCount(): Long

  def offer(a: A): Boolean

  def unsafePoll(): B // returns null when queue empty
  def poll(): Option[B] = Option(unsafePoll())
}
```

Rough plan:

1. Write a straightforward but correct implementation based on "[A Portable Lock-Free Bounded Queue](http://pirkelbauer.com/papers/icapp16.pdf)".
2. Write benchmarks and establish baseline performance. Measuring performance here could be a bit tricky since interesting benchmarks needs to be multithreaded, asynchronous and the system should be in a stable state during the iteration.
3. Iteratively optimize performance. Things to consider here:
    1. Using `null`s instead of Options for better performance.
    2. Padding fields to avoid false sharing.
    3. Using a power of two capacity to make index calculations faster.
    4. ???

**Details**

There are a couple approaches to designing a lock-free concurrent queue (CoralQueue, LMAX, JCTools, etc.). One difference is in the way elements are added to the queue.

First approach (LMAX, CoralQueue) looks like this:

```scala
abstract class RingBuffer[A](val size: Int) {
  def nextSequence(): Option[Long]
  def publish(seq: Long): Unit
  def poll(): Option[A]
}
val q = new RingBuffer(???) {...}

// Thread 1
val reservedSlot = q.nextSequence().get // ignore that queue could be full and we'd need to retry
val el = ...
q(reservedSlot) = el
q.publish(reservedSlot)

// Thread 2
val nextElOpt = q.poll()
```

The phases are: reserve a place in the queue, put an element in the queue and publish it when ready.

The big upside of this approach is **zero garbage** when you deal with primitive queue elements or mutable objects comprised of primitives and can pre-allocate queue elements.

The downside is publishing items is more involved. Consider two threads. Thread 1 and 2 reserved sequence numbers 1 and 2 respectively. Assume they both put their elements into the queue, but before Thread 1 managed to publish it got descheduled. Thread 2 cannot publish before Thread 1 publishes, so it has to wait. There are ways to avoid spinning (e.g. LMAX v3), but nevertheless extra complexity on the caller's side exists.

Second approach (JCTools, Java's `ConcurrentLinkedQueue`) is more standard:

```scala
abstract class RingBuffer[A] {
  def offer(a: A): Boolean
  def poll(): Option[B]
}
val q = new RingBuffer(???) {...}

// Thread 1
val el = ...
q.offer(el)

// Thread 2
val nextElOpt = q.poll()
```

In this case `offer` and `poll` are atomic operations on their own and there's no extra complexity on the caller's side. It plays nicely with immutable messages being passed through the queue.

JCTools has an excellent Java implementation of MPMC queue based on a modified design by D. Vyukov (the original design doesn't have all the lock-free properties, like termination safety and resistance to priority inversion, but is simple and fast).

In terms of interface I prefer the second approach. As for the implementation I plan to use the design described in "[A Portable Lock-Free Bounded Queue](http://pirkelbauer.com/papers/icapp16.pdf)" [LFBQ] by Pirkelbauer et al. which does have the aforementioned lock-free properties. However, the design is for C++11 memory model, so it's not 100% clear yet whether it can be fully adapted for JVM.

**RTS**
I didn't think too much about how `RTS` could be using a ring buffer under the hood, but my current feeling is that for every capacity it's possible to construct an IO that would block all the slots in the queue and make further progress impossible. Might not be a real problem though...

**Comment by jdegoes** [comment]
@artempyanykh Well thought out plan. 😄

**Comment by regiskuckaertz** [comment]
🎉 ✨ 🚀
